src/Analyzers/Core/Analyzers/UseCollectionInitializer/AbstractObjectCreationExpressionAnalyzer.cs

```diff
@@ -2,137 +2,138 @@
+    protected void Clear()
+    {
+        State = default;
+        _objectCreationExpression = null!;
+        _analyzeForCollectionExpression = false;
+    }
 
-        protected ISyntaxFacts SyntaxFacts => this.State.SyntaxFacts;
-        protected SemanticModel SemanticModel => this.State.SemanticModel;
+    protected ImmutableArray<TMatch> AnalyzeWorker(CancellationToken cancellationToken)
+    {
+        if (!ShouldAnalyze(cancellationToken))
+            return default;
 
-        protected abstract bool ShouldAnalyze(CancellationToken cancellationToken);
-        protected abstract bool TryAddMatches(ArrayBuilder<TMatch> matches, CancellationToken cancellationToken);
+        using var _ = ArrayBuilder<TMatch>.GetInstance(out var matches);
+        if (!TryAddMatches(matches, cancellationToken))
+            return default;
 
-        public void Initialize(
-            UpdateExpressionState<TExpressionSyntax, TStatementSyntax> state,
-            TObjectCreationExpressionSyntax objectCreationExpression,
-            bool analyzeForCollectionExpression)
-        {
-            State = state;
-            _objectCreationExpression = objectCreationExpression;
-            _analyzeForCollectionExpression = analyzeForCollectionExpression;
-        }
+        return matches.ToImmutable();
+    }
 
-        protected void Clear()
-        {
-            State = default;
-            _objectCreationExpression = null;
-            _analyzeForCollectionExpression = false;
-        }
+    protected UpdateExpressionState<TExpressionSyntax, TStatementSyntax>? TryInitializeState(
+        SemanticModel semanticModel,
+        ISyntaxFacts syntaxFacts,
+        TObjectCreationExpressionSyntax rootExpression,
+        CancellationToken cancellationToken)
+    {
+        var statement = rootExpression.FirstAncestorOrSelf<TStatementSyntax>()!;
+        if (statement is null)
+            return null;
 
-        protected ImmutableArray<TMatch> AnalyzeWorker(CancellationToken cancellationToken)
-        {
-            if (!ShouldAnalyze(cancellationToken))
-                return default;
+        return
+            TryInitializeVariableDeclarationCase(semanticModel, syntaxFacts, rootExpression, statement, cancellationToken) ??
+            TryInitializeAssignmentCase(semanticModel, syntaxFacts, rootExpression, statement, cancellationToken);
+    }
 
-            using var _ = ArrayBuilder<TMatch>.GetInstance(out var matches);
-            if (!TryAddMatches(matches, cancellationToken))
-                return default;
+    private UpdateExpressionState<TExpressionSyntax, TStatementSyntax>? TryInitializeVariableDeclarationCase(
+        SemanticModel semanticModel,
+        ISyntaxFacts syntaxFacts,
+        TObjectCreationExpressionSyntax rootExpression,
+        TStatementSyntax containingStatement,
+        CancellationToken cancellationToken)
+    {
+        if (containingStatement is not TLocalDeclarationStatementSyntax localDeclarationStatement)
```

Comment:
this function update is the meat of the change.  Previously it made assumptions about the shape that didn't hold in VB.  now it calls into each language to deal with the specific shapes, and only uses shared code once it knows it can.